### PR TITLE
Use shallow clones when fetching repositories

### PR DESCRIPTION
**Pull request by DanielTomlinson**

### CODE REVIEW

**Comment by orta** [comment]
interesting, @CocoaPods/core / anyone know if it's possible to continue development from a shallow clone? e.g. adding new commits etc?

**Comment by segiddins** [comment]
It is with a sufficiently recent git

**Comment by orta** [comment]
Then this seems :+1:  to me

File: Rakefile [hunk 1]
```diff
@@ -547,7 +547,7 @@ def clone_repos(repos)
     if File.exist?(name)
       puts "Already cloned"
     else
-      sh "git clone #{url}"
+      sh "git clone #{url} --depth 1"
```
Comment (segiddins) [inline Rakefile]: might also want to slap a `--recursive` on there as well

**Comment by DanielTomlinson** [comment]
@segiddins Updated :+1: 

@orta :sparkles:

**Comment by manuyavuz** [comment]
It feels like a really good update 👍

**Comment by kylef** [comment]
This would render git-blame and git-bisect (any various other git tools) useless as local checkout's won't have any diff information. To me this change doesn't seem ideal.

**Comment by orta** [comment]
Could add another command to transform them to deep clones that could be mentioned at the end of the rake clone task?

**Comment by DanielTomlinson** [comment]
@orta Yeah, `git fetch --unshallow` in recent git versions will handle that quite nicely.

**Comment by manuyavuz** [comment]
It could be better to clone repos shallow initially because rainforest puts several repos together and not every people work simultaneously with each of the repos. @kylef I really understand your concerns, but imo providing deep clone as an option could be better because it would be very easy to switch to a deep local copy when needed.

**Comment by neonichu** [comment]
I would agree with @kylef here, but it seems fine if the default `rake clone` does shallow as long as we would have a non-shallow task as well — I always want to have the non-shallow clone of everything and I'm thinking there are at least a few others who want the same, so it seems annoying manually having to unshallow every single repo.

**Comment by manuyavuz** [comment]
Agree on having a deep option which does full clone on each repo. I also like to use repos with full history, but it felt to ease the process for people who want to contribute with PRs on particular repos.

**Comment by DanielTomlinson** [comment]
So also add an argument to `bootstrap` to enable deep clones for those who want them?

**Comment by neonichu** [comment]
@DanielTomlinson sounds good :+1:

**Comment by kylef** [comment]
Providing there is now the `git fetch --unshallow` option I'm fine with these changes :+1:.

**Comment by orta** [comment]
Alright, lets get a unshallow flag added and this PR shipped :dancers:

**Comment by DanielTomlinson** [comment]
:+1:, I'll finish this off tomorrow afternoon.

**Comment by DanielTomlinson** [comment]
@orta Done.

**Comment by orta** [comment]
Awesome beans - this travis fail isn't related.

**Comment by kylef** [comment]
Just want to mention that `--unshallow` isn't complete and more steps may be required to convert to a non-shallow checkout.

Using both `git fetch --unshallow`, `git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"` and `git fetch` I managed to convert back:

```shell
$ git l
* c38786f - (grafted, HEAD, tag: 0.2.0) Nest 0.2.0 (3 months ago) <Kyle Fuller>

$ git fetch --unshallow
remote: Counting objects: 45, done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 45 (delta 27), reused 43 (delta 25), pack-reused 0
Unpacking objects: 100% (45/45), done.
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
From https://github.com/nestproject/Nest
 * [new tag]         0.1.0      -> 0.1.0

$ git checkout master
error: pathspec 'master' did not match any file(s) known to git.

$ git checkout 0.2.1
error: pathspec '0.2.1' did not match any file(s) known to git.

$ git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
$ git fetch
remote: Counting objects: 27, done.
remote: Total 27 (delta 7), reused 7 (delta 7), pack-reused 20
Unpacking objects: 100% (27/27), done.
From https://github.com/nestproject/Nest
 * [new branch]      master     -> origin/master
 * [new tag]         0.2.1      -> 0.2.1
```

**Pull request merged**